CohortExtractorError appears to be pickled incorrectly

When the database connection string is incorrect (and potentially with other errors too), the step that runs docker raises an exception.

However, we get the following traceback, which implies that `pebble` (which we're using for its nice async API) is failing to pickle the exception correctly: 

```
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/threading.py", line 890, in _bootstrap
job-runner_1  |     self._bootstrap_inner()
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/threading.py", line 932, in _bootstrap_inner
job-runner_1  |     self.run()
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/threading.py", line 870, in run
job-runner_1  |     self._target(*self._args, **self._kwargs)
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/site-packages/pebble/pool/process.py", line 178, in message_manager_loop
job-runner_1  |     pool_manager.process_next_message(SLEEP_UNIT)
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/site-packages/pebble/pool/process.py", line 210, in process_next_message
job-runner_1  |     message = self.worker_manager.receive(timeout)
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/site-packages/pebble/pool/process.py", line 342, in receive
job-runner_1  |     return self.pool_channel.recv()
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/site-packages/pebble/pool/channel.py", line 63, in recv
job-runner_1  |     return self.reader.recv()
job-runner_1  |   File "/root/.pyenv/versions/3.8.3/lib/python3.8/multiprocessing/connection.py", line 251, in recv
job-runner_1  |     return _ForkingPickler.loads(buf.getbuffer())
File "/home/seb/Code/opensafely/job-runner/runner/exceptions.py", line 22, in __init__
    assert (
AssertionError: `report_args` keyword arg must be supplied to CohortExtractorError
```
```

--- a/runner/exceptions.py
+++ b/runner/exceptions.py
@@ -8,22 +8,10 @@ people outside the secure platform.
 """
 
 
-class ErrorWithKeywordFactory:
-    def __new__(self, *args):
-        klass = args[0]
-        report_args = args[1]
-        args = args[1:]
-        return klass(*args, report_args=report_args)
-
-
 class OpenSafelyError(Exception):
     safe_args = False
 
-    def __init__(self, *args, report_args=None):
-        classname = type(self).__name__
-        assert (
-            report_args is not None
-        ), f"`report_args` keyword arg must be supplied to {classname}"
+    def __init__(self, *args, report_args=False):
         self.report_args = report_args
         assert self.status_code not in [-1, 99], "status_codes -1 and 99 are reserved"
         super().__init__(*args)
@@ -34,22 +22,6 @@ class OpenSafelyError(Exception):
             return classname + ": " + " ".join(self.args)
         else:
             return classname + ": [possibly-unsafe details redacted]"
-
-    def __reduce__(self):
-        """Override BaseException's custom pickling implementation, to support
-        keyword arguments.
-        """
-        # The default implementation instantiates an instance with
-        # just positional arguments, *then* updates its __dict__ with
-        # keyword arguments.  This breaks the validation in this class'
-        # __init__ function.
-        #
-        #
-        # See https://stackoverflow.com/a/41809333/559140 for context
-        return (
-            ErrorWithKeywordFactory,
-            (self.__class__, self.report_args) + self.args,
-        )
 
 
 class DockerError(OpenSafelyError):

--- a/runner/exceptions.py
+++ b/runner/exceptions.py
@@ -8,14 +8,25 @@ people outside the secure platform.
 """
 
 
+class ErrorWithKeywordFactory:
+    def __new__(self, *args):
+        klass = args[0]
+        report_args = args[1]
+        args = args[1:]
+        return klass(*args, report_args=report_args)
+
+
 class OpenSafelyError(Exception):
     safe_args = False
 
-    def __init__(self, *args, report_args=None, **kwargs):
-        assert report_args is not None, "`report_args` keyword arg must be supplied"
+    def __init__(self, *args, report_args=None):
+        classname = type(self).__name__
+        assert (
+            report_args is not None
+        ), f"`report_args` keyword arg must be supplied to {classname}"
         self.report_args = report_args
         assert self.status_code not in [-1, 99], "status_codes -1 and 99 are reserved"
-        super().__init__(*args, **kwargs)
+        super().__init__(*args)
 
     def safe_details(self):
         classname = type(self).__name__
@@ -23,6 +34,22 @@ class OpenSafelyError(Exception):
             return classname + ": " + " ".join(self.args)
         else:
             return classname + ": [possibly-unsafe details redacted]"
+
+    def __reduce__(self):
+        """Override BaseException's custom pickling implementation, to support
+        keyword arguments.
+        """
+        # The default implementation instantiates an instance with
+        # just positional arguments, *then* updates its __dict__ with
+        # keyword arguments.  This breaks the validation in this class'
+        # __init__ function.
+        #
+        #
+        # See https://stackoverflow.com/a/41809333/559140 for context
+        return (
+            ErrorWithKeywordFactory,
+            (self.__class__, self.report_args) + self.args,
+        )
 
 
 class DockerError(OpenSafelyError):

--- a/tests/test_runner.py
+++ b/tests/test_runner.py
@@ -1,15 +1,19 @@
 import requests_mock
 import runner
 from unittest.mock import patch
+import pickle
 import time
 
 from runner import make_volume_name
 from runner import make_container_name
-from runner.exceptions import CohortExtractorError
 from runner.exceptions import OpenSafelyError
 from runner.exceptions import RepoNotFound
 
 import pytest
+
+
+class TestError(OpenSafelyError):
+    status_code = 10
 
 
 @pytest.fixture(scope="function")
@@ -111,9 +115,6 @@ def test_bad_volume_name_raises():
 
 
 def test_exception_reporting():
-    class TestError(OpenSafelyError):
-        status_code = 10
-
     error = TestError("thing not to leak", report_args=False)
     assert error.safe_details() == "TestError: [possibly-unsafe details redacted]"
     assert repr(error) == "TestError('thing not to leak')"
@@ -133,3 +134,10 @@ def test_reserved_exception():
 
     with pytest.raises(RepoNotFound):
         raise RepoNotFound(report_args=True)
+
+
+def test_pickled_exception():
+    error = TestError(report_args=True)
+    pickled_error = pickle.dumps(error)
+    unpickled_error = pickle.loads(pickled_error)
+    assert error.__dict__ == unpickled_error.__dict__

--- a/tests/test_runner.py
+++ b/tests/test_runner.py
@@ -134,10 +134,3 @@ def test_reserved_exception():
 
     with pytest.raises(RepoNotFound):
         raise RepoNotFound(report_args=True)
-
-
-def test_pickled_exception():
-    error = TestError(report_args=True)
-    pickled_error = pickle.dumps(error)
-    unpickled_error = pickle.loads(pickled_error)
-    assert error.__dict__ == unpickled_error.__dict__

